RISC-V header cleanups

gmarull opened the pull request.

gmarull committed:
soc: riscv: cleanup usage/definition of MCAUSE IRQ flag

The MCAUSE register has the "Interrupt" flag defined defined at XLEN-1
position (31 for 32-bit, 63 for 64-bit). This is not an SoC specific
option, and there's no need to expose it publicly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: riscv: define some RISC-V exception codes

As defined in Table 3.6 of "The RISC-V Instruction Set Manual, Volume
II: Privileged Architecture". Delete all spread definitions of the same,
weirdly prefixed with "SOC".

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv: gd32vf103: simplify MCAUSE exception mask handling

The exception mask needs to cover MCAUSE bits 11:0, there's no need to
overengineer this setting using DT properties.

Ref. https://doc.nucleisys.com/nuclei_spec/isa/core_csr.html#mcause

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: riscv: remove SOC from RISCV_SOC_MCAUSE_EXCEPTION_MASK

Just to stay consistent with other RISC-V related settings.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: riscv: use CONFIG_RISCV_MCAUSE_EXCEPTION_MASK

Instead of custom SOC_MCAUSE_EXP_MASK definition. Note that SoCs
selecting RISCV_PRIVILEGED already used such config indirectly (see
changes in soc_common.h).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: riscv: define RISC_IRQ_MSOFT/MEXT

Instead of relying on spread definitions within SoC files.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
arch: riscv: define local soc_interrupt_init prototypes

Instead of relying on messy soc.h files which are included via a fragile
chain of includes.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv: riscv-privileged: drop soc_common.h

The header file is no longer needed.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

carlocaione commented on include/zephyr/arch/riscv/irq.h:
```diff
@@ -32,6 +32,12 @@ extern "C" {
 /** Environment Call from M-mode */
 #define RISCV_EXC_ECALLM 11
 
+#ifdef CONFIG_64BIT
+#define RISCV_MCAUSE_IRQ_BIT          (1 << 63)
+#else
+#define RISCV_MCAUSE_IRQ_BIT          (1 << 31)
+#endif
```
yeeeeaahh, we both now at least one platform where this is not true ...

gmarull committed:
arch: riscv: irq_manage: add missing PLIC header

Module uses RISC-V PLIC interrupt controller API without including the
necessary headers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: gpio: sifive: add missing PLIC header

Driver is using the PLIC API without including the right header.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: gpio: b91: add missing PLIC header

Driver is using the RISC-V PLIC interrupt controller without including
the necessary headers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: ieee802154: b91: add missing PLIC include

Driver was using RISC-V PLIC API without including the necessary
headers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: serial: b91: add missing RISC-V PLIC header

Driver was using PLIC without including necessary headers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

fkokosinski reviewed (approved):
LGTM

ycsin commented:
I'm doing an update to v3.7 and my application doesn't build, partially due to the changes in this PR. 

Given that the `soc.h` is exposed to the application/driver/subsys, can we provide the old macros removed/renamed in this PR and use the deprecation process?

cc @carlocaione @gmarull

gmarull commented:
> I'm doing an update to v3.7 and my application doesn't build, partially due to the changes in this PR.
> 
> Given that the `soc.h` is exposed to the application/driver/subsys, can we provide the old macros removed/renamed in this PR and use the deprecation process?
> 
> cc @carlocaione @gmarull

I don't think the content of soc.h can be considered "stable", soc.h has been largely removed as it introduces all sorts of problems.

ycsin commented:
> I don't think the content of soc.h can be considered "stable"

Sure, but these common macros have been there for years and exposed to ..everything, it isn't fun to find the applications fail to compile in a sea of compilation error and have to dig through the versions/commits/PRs to find where they are.

gmarull commented:
> > I don't think the content of soc.h can be considered "stable"
> 
> Sure, but these common macros have been there for years and exposed to ..everything, it isn't fun to find the applications fail to compile in a sea of compilation error and have to dig through the versions/commits/PRs to find where they are.

Applications should not rely on soc.h ;-)

ycsin commented:
> Applications should not rely on soc.h ;-)

I meant the whole thing really. Application doesn't compile if SOC/boards/drivers/subsys/lib doesn't compile ¯\\_(ツ)_/¯ The changes here can break OOT stuff that has something to do with interrupt/exception, i.e. PLIC driver

It becomes much more complicated when the OOT stuff is shared by multiple boards targeting different Zephyr versions 🥲

And meanwhile `soc.h` is still being included in our samples:

```
./zephyr/samples/drivers/kscan/src/main.c:10
./zephyr/samples/drivers/ps2/src/main.c:10
./zephyr/samples/drivers/led_xec/src/main.c:8
./zephyr/samples/drivers/clock_control_xec/src/main.c:8
./zephyr/samples/drivers/clock_control_xec/src/main.c:18
./zephyr/samples/drivers/espi/src/espi_oob_handler.c:8
./zephyr/samples/drivers/espi/src/main.c:10
./zephyr/samples/drivers/peci/src/main.c:10
./zephyr/samples/bluetooth/mesh_demo/src/microbit.c:10
./zephyr/samples/boards/mec172xevb_assy6906/rom_api/src/main.c:17
./zephyr/samples/boards/mec172xevb_assy6906/qmspi_ldma/src/main.c:10
./zephyr/samples/boards/esp32/flash_memory_mapped/src/main.c:14
./zephyr/samples/boards/esp32/flash_encryption/src/main.c:15
./zephyr/samples/boards/mec15xxevb_assy6853/power_management/src/power_mgmt.c:10
```

and a lot of other places of course..
